setup/helpers/abs_helper.rb

```diff
@@ -0,0 +1,657 @@
+  def backoff_sleep(tries)
+    # Exponential with some randomization
+    sleep_time = 2**tries
+    puts "Sleeping for #{sleep_time} seconds after attempt #{tries}..."
+    sleep sleep_time
+    nil
+  end
+
+  # Verifies that the specified host can be accessed via ssh as the root user
+  #
+  # @author Bill Claytor
+  #
+  # @param [string] host The host to verify
+  #
+  # @return [true, false] Based on successful verification of the specified host
+  #
+  # @example
+  #   success = abs_verify_host(host)
+  #
+  def verify_abs_host(host)
+    result = nil
+    success = false
+    user = "root"
+    key_file = "~/.ssh/abs-aws-ec2.rsa"
+
+    puts "Verifying #{host}"
+    puts
+
+    # TODO: 8 tries?
+    for tries in 1..8
+
+      puts "Attempt #{tries} for #{host}"
+
+      begin
+        # verify that Beaker will be able to start using the host
+        ssh = Net::SSH.start(host, user, keys: key_file)
+        result = ssh.exec!("rpm -q curl")
+        ssh.close
+
+        if result
+          puts "Result: #{result}"
+          raise "Error: root account is not yet configured" if result.to_s.include?("centos")
+          success = true
+          break
+
+        else
+          raise "Unknown error"
+        end
+
+      rescue => err
+        puts "Attempted connection to #{host} failed with '#{err}'"
+        backoff_sleep(tries)
+
+      end
+
+    end
+
+    puts "Failed to verify host #{host}" unless success
+    return success
+  end
+
+  # Verifies that the specified hosts can be accessed via ssh as the root user
+  #
+  # @author Bill Claytor
+  #
+  # @param [Array] hosts The hosts to verify
+  #
+  # @return [true, false] Based on successful verification of the specified hosts
+  #
+  # @example
+  #   success = abs_verify_hosts(host)
+  #
+  def verify_abs_hosts(hosts)
+    success = false
+    puts "Verifying ABS hosts: #{hosts}"
+    hosts.each do |host|
+      puts
+      puts "Current host: #{host}"
+
+      success = verify_abs_host(host["hostname"])
+      break unless success
+    end
+
+    puts "Unable to verify the provisioned hosts" unless success
+    return success
+  end
+
+  # Determines whether the specified hosts list is valid
+  #
+  # @author Bill Claytor
+  #
+  # @param [JSON] abs_resource_hosts The hosts to validate
+  #
+  # @return [true, false] Based on the validity of the specified hosts
+  #
+  # @example
+  #   valid = abs_valid_resource_hosts?(abs_resource_hosts)
+  #
+  def valid_abs_resource_hosts?(abs_resource_hosts)
+    is_valid = false
+
+    if abs_resource_hosts.nil?
+      puts "A valid hosts array is required; nil was specified"
+      puts
+    else
+
+      begin
+        hosts = JSON.parse(abs_resource_hosts)
+        host = hosts[0]
+        hostname = host["hostname"]
+        if !hostname.nil? && !hostname.empty?
+          is_valid = true
+        else
+          puts "The specified resource host array is not valid: #{abs_resource_hosts}"
+          puts
+        end
+      rescue
```

Comment:
should be: 
rescue JSON::ParseError => ex
and we should re-raise so we have this context if it ever fails.

Reply:
Removed